### DISCUSSION TITLE
OSC in General

### DISCUSSION THREAD

**Discussion by suce-ex** [RomanKubiak/ctrlr#461] [General]
Hi, 
I figured out OSC works pretty well on Standalone (CTRLR v5.3.201) in both directions : 
- SEND
- RECEIVE

However as panel in VST (in Cubase) (and not as "exported Instance") it reacts like this :
-Sending out correctly (from panel)
-Receive: IT DOESNT RECEIVE anymore (after a while... couple of minutes)

It receives AGAIN, if an OSC msg is SENT (from panel). But ... later on (minutes later) ... looses connection... 
(please remember, it does work perfectly in Standalone (CTRLR v5.3.201))

Anyone?
s-h

**Comment by suce-ex** [comment]
for those who wonder:

-ctrlr (v5.3.201) works perfectly well for arguments[0].f as mostly used in TouchOSC. 

receiving osc is:
if path == path and arguments [0].f == .. then

sending is :
osc.createMessage ("path", "1 or f or i ") 

there were complainings in the forum in the past about other arguments "s" ... however argument f (float) works pretty well. Brickwall.

**Comment by suce-ex** [comment]
I sent out a pseudo timer, which sends out osc messages every minute. And that works so far. [Connection stable]
However, it's a lot of timer messages for 30 minutes. 
How to set a timer which loops every 1 minute ?

**Comment by damiensellier** [comment]
Hi @suce-ex , that's pretty simple, here is a procedure to start a timer loop.

Trigger the timer wherever you want :
```
	timer:setCallback (1, TimerCallback)
	timer:startTimer(1, 60000) -- timerID 1 for 60*1000ms cycles

```


Create a TimerCallback() LUA method :

```
TimerCallback = function(timerId)
-- Timer to send a message every 01'00"
	if timerId == 1 then
		-- send your message here
	end

end -- f

```

PS: don't worry too much about the amount of message sent if your external device is modern, think about how a midi time clock is sent and manage during a minute...

**Reply by suce-ex** [reply]
thanks so far @damiensellier , 
but isnt smth missing? i dont see how the timer would loop in your code

**Reply by suce-ex** [reply]
wait a sec i think i got you

**Reply by damiensellier** [reply]
A timer() is by essence a JUCE function that runs operations in cycle within a predefined period.
timer:startTimer(1, 60000) -- start the timer loop with Index 1, that is going to loop every 60sec, and is defined in the TimerCallback() function

**Reply by suce-ex** [reply]
i got it. and it WORKS for a constant OSC connection!  (tested it on hours, stays stable even if the client disconnects (!))

(I thought timers must end by "timer:stopTimer (timerId) . That made me confused. That's why I wasn't sure how your code works. )

To be sure that this clock is the reason OSC stands stable, and for educational matters:

could you please make me understand how to make the timer starting on a button, (toggle-button or just button)
where: toggled on > the timer starts and: toggled off > the timer goes into "timer:stopTimer (timerId) 
?!

[and is this also possible WITHOUT using the integrated "called when the modulator value changes" ?]

Thanks so much

**Reply by damiensellier** [reply]
Sure that's very easy.
You have to add a button on your template.
Then you have to assign a script triggered when the modulator value changes (0->1 or 1->0).
This script check the new modulator value
if it's "1" then it starts the timer, if it's "0" it stops the timer.

This is the way it works. I'll post it as a demo panel in 5 minutes ;)

**Reply by damiensellier** [reply]
Here is the demo panel :
[TimerSwitch_1_0_0_2022-05-03_18-17.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8614580/TimerSwitch_1_0_0_2022-05-03_18-17.panel.zip)

NOTE: I noticed the command within timers always starts at the end of the period. for exemple, if you set your timer for a 1min cycle, you'll get your command sent 1 minute after you clicked the button. Not at time 0 but at time +1min. SO if you need it your command to be sent right away just add the command before the call to start the timer.

**Reply by suce-ex** [reply]
got it! Thanks so much. Didnt know that "value" is somehow global or index. that actually confuses me. 

getting back to question B: is such a message possible without being integrated into the panel "called when the modulator value changes" ?

**Comment by damiensellier** [comment]
Nope you can't trigger the timer without an event. (Mod value change, click, double click, ...)
OR, if you just need to run this script from start to end of your session, you can put the timer trigger into the global method "when the panel has finished loading". This way the timer will start as soon as you open the panel and is ready to go.

**Reply by damiensellier** [reply]
Here is the demo panel with the autostart when the panel is ready :
[TimerSwitch_1_0_1_2022-05-04_12-37.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8622335/TimerSwitch_1_0_1_2022-05-04_12-37.panel.zip)

The method is loaded in the global parameters
<img width="558" alt="Capture d’écran 2022-05-04 à 12 36 18" src="https://user-images.githubusercontent.com/55164385/166729360-33cbd70c-8af4-4cf2-8ecc-135b8dd965c4.png">

**Reply by suce-ex** [reply]
Works brickwall! 
For the community, which is interested into OSC: 
- This "ping out timer" to Host keeps the connection from Client to Host stable!

Thanks to your exemples @damiensellier ,  I can surely say, now, that this is what keeps OSC in both ways (host/client) connected! (at least in version CTRLR v5.3.201)

Excellent, hope people come on board with this knowledge in head. <3


[BTW, I can feel my old phone heatening up with 6sec of pings. So this is also a great functionality in winter! (as well an incredible way to lower the battery, just observing) >>> I'll check how many seconds or minutes it would actually need to stay connected. :)]

**Reply by suce-ex** [reply]
So why did it work brickwall in Standalone, 
but not as VST without that ping? (to be honest, i dont care, since it works that way now. been curious.)

**Reply by damiensellier** [reply]
I have no idea really
